test/integration/client/exec_test.go

```diff
@@ -537,6 +540,138 @@ func TestExecPluginViaInformer(t *testing.T) {
+	time.Sleep(lifetime)
+}
+
+func (e *execPlugin) rotateCert(cert, key string, lifetime time.Duration) {
+	e.t.Helper()
+
+	expirationTimestamp := metav1.NewTime(time.Now().Add(lifetime)).Format(time.RFC3339Nano)
+	newOutput := fmt.Sprintf(`{
+		"kind": "ExecCredential",
+		"apiVersion": "client.authentication.k8s.io/v1beta1",
+		"status": {
+			"expirationTimestamp": %q,
+			"clientCertificateData": %q,
+			"clientKeyData": %q
+		}
+	}`, expirationTimestamp, cert, key)
+	if err := os.WriteFile(e.outputFile.Name(), []byte(newOutput), 0644); err != nil {
+		e.t.Fatal(err)
+	}
+
+	// Wait for old token to expire plus some fudge room.
+	time.Sleep(lifetime)
+}
+
+func TestExecPluginRotationViaInformer(t *testing.T) {
+	result, clientAuthorizedToken, _, _ := startTestServer(t)
+	const clientUnauthorizedToken = "invalid-token"
+	const tokenLifetime = time.Second * 5
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*5)
+	defer cancel()
+
+	adminClient := clientset.NewForConfigOrDie(result.ClientConfig)
+	ns := createNamespace(ctx, t, adminClient)
+
+	clientDialer := connrotation.NewDialer((&net.Dialer{
+		Timeout:   30 * time.Second,
+		KeepAlive: 30 * time.Second,
+	}).DialContext)
+
+	execPlugin := newExecPlugin(t)
+
+	clientConfig := rest.AnonymousClientConfig(result.ClientConfig)
+	clientConfig.ExecProvider = execPlugin.config()
+	clientConfig.Dial = clientDialer.DialContext
+	clientConfig.Wrap(func(rt http.RoundTripper) http.RoundTripper {
+		// This makes it helpful to see what is happening with the informer's client.
+		return transport.NewDebuggingRoundTripper(rt, transport.DebugCurlCommand, transport.DebugURLTiming)
+	})
+
+	// Initialize informer spy wth invalid token.
+	// Make sure informer never syncs because it can't authenticate.
+	execPlugin.rotateToken(clientUnauthorizedToken, tokenLifetime)
+	informer, informerSpy := startConfigMapInformer(ctx, t, clientset.NewForConfigOrDie(clientConfig), ns.Name)
+	waitForInformerSync(t, informer, false)
+	createUpdateDeleteConfigMap(ctx, t, adminClient.CoreV1().ConfigMaps(ns.Name))
+	informerSpy.waitForEvents(t, false)
+
+	// Rotate token to valid token.
+	// Make sure informer sees events because it now has a valid token with which it can authenticate.
+	execPlugin.rotateToken(clientAuthorizedToken, tokenLifetime)
+	waitForInformerSync(t, informer, true)
+	informerSpy.clear()
+	createdCM, updatedCM, deletedCM := createUpdateDeleteConfigMap(ctx, t, adminClient.CoreV1().ConfigMaps(ns.Name))
+	informerSpy.waitForEvents(t, true)
+	assertInformerEvents(t, informerSpy, createdCM, updatedCM, deletedCM)
+
+	// Rotate token to something invalid and clip watch connection.
+	// Informer should recreate connection with invalid token.
+	// Make sure informer does not see events since it is using the invalid token.
+	execPlugin.rotateToken(clientUnauthorizedToken, tokenLifetime)
+	clientDialer.CloseAll()
+	time.Sleep(time.Second * 60) // Wait for informer to be up
```

Comment:
Is there a better way to assert this without waiting 60 seconds? I thought about checking the `informer.GetController().LastSyncResourceVersion()` but I don't know if that would help me because the informer has already synced above. I supposed I could create a new `ConfigMap` and wait for the controller to be up to date with that resource version?